File: backend/payment/src/main/java/com/bbmovie/payment/service/momo/MomoAdapter.java
package com.bbmovie.payment.service.momo;

import com.bbmovie.payment.dto.request.PaymentRequest;
import com.bbmovie.payment.dto.response.PaymentCreationResponse;
import com.bbmovie.payment.dto.response.PaymentVerificationResponse;
import com.bbmovie.payment.dto.response.RefundResponse;
import com.bbmovie.payment.entity.enums.PaymentProvider;
import com.bbmovie.payment.entity.enums.PaymentStatus;
import com.bbmovie.payment.service.PaymentProviderAdapter;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.Base64;

/**
 * MOMO banned the developer account
 */
@Slf4j
@Service("momo")
public class MomoAdapter implements PaymentProviderAdapter {

    @Value("${payment.momo.partner-code}")
    private String partnerCode;

    @Value("${payment.momo.access-key}")
    private String accessKey;

    @Value("${payment.momo.secret-key}")
    private String secretKey;

    @Value("${payment.momo.sandbox:true}")
    private boolean sandbox;

    @Value("${payment.momo.redirect-url}")
    private String redirectUrl;

    @Value("${payment.momo.ipn-url}")
    private String ipnUrl;

    private final ObjectMapper mapper = new ObjectMapper();

    private final boolean supported = false;

    @Override
    public PaymentCreationResponse processPayment(PaymentRequest request, HttpServletRequest httpServletRequest) {
        if (!supported) {
            throw new UnsupportedOperationException("Momo is not supported yet");
        }

        long amount = Optional.ofNullable(request.getAmount())
                .map(BigDecimal::longValue)
                .orElseThrow(() -> new IllegalArgumentException("amount is required"));
        String orderId = Optional.ofNullable(request.getOrderId())
                .orElse("Partner_Transaction_ID_" + System.currentTimeMillis());
        String requestId = "Request_ID_" + System.currentTimeMillis();
        String orderInfo = "Payment for order " + orderId;
        String requestType = MomoConstraint.REQUEST_TYPE_CAPTURE_WALLET;

        String extraDataJson;
        try {
            extraDataJson = mapper.writeValueAsString(Map.of("skus", ""));
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize extraData", e);
        }
        String extraData = Base64.getEncoder().encodeToString(extraDataJson.getBytes(StandardCharsets.UTF_8));

        Map<String, Object> body = new HashMap<>();
        body.put("partnerCode", partnerCode);
        body.put("requestId", requestId);
        body.put("amount", amount);
        body.put("orderId", orderId);
        body.put("orderInfo", orderInfo);
        body.put("redirectUrl", redirectUrl);
        body.put("ipnUrl", ipnUrl);
        body.put("requestType", requestType);
        body.put("extraData", extraData);

        String rawSignature = buildSignatureString(Map.of(
                "accessKey", accessKey,
                "amount", String.valueOf(amount),
                "extraData", extraData,
                "ipnUrl", ipnUrl,
                "orderId", orderId,
                "orderInfo", orderInfo,
                "partnerCode", partnerCode,
                "redirectUrl", redirectUrl,
                "requestId", requestId,
                "requestType", requestType
        ));

        String signature = hmacSha256Hex(secretKey, rawSignature);
        body.put("signature", signature);

        String url = sandbox ? MomoConstraint.CREATE_URL_TEST : MomoConstraint.CREATE_URL_PROD;
        String responseBody = sendJson(url, body);


        try {
            @SuppressWarnings("unchecked")
            Map<String, Object> response = mapper.readValue(responseBody, Map.class);

            int resultCode = Optional.ofNullable((Number) response.get("resultCode")).map(Number::intValue).orElse(-1);
            String payUrl = Optional.ofNullable(response.get("payUrl")).map(Object::toString).orElse(null);

            PaymentStatus status = (resultCode == 0) ? PaymentStatus.PENDING : PaymentStatus.FAILED;

            return PaymentCreationResponse.builder()
                    .transactionId(orderId)
                    .status(status)
                    .providerReference(payUrl)
                    .build();
        } catch (Exception e) {
            log.error("Failed to process payment: {}", e.getMessage());
            return PaymentCreationResponse.builder()
                    .transactionId(null)
                    .status(null)
                    .providerReference(null)
                    .build();
        }
    }

    @Override
    public PaymentVerificationResponse verifyPaymentCallback(Map<String, String> paymentData, HttpServletRequest httpServletRequest) {
        if (!supported) {
            throw new UnsupportedOperationException("Momo is not supported yet");
        }

        String signature = paymentData.get("signature");
        Integer resultCode = Optional.ofNullable(paymentData.get("resultCode")).map(Integer::parseInt).orElse(null);
        if (signature == null || signature.isBlank() || resultCode == null) {
            return new PaymentVerificationResponse(false, null, null, null, null, null);
        }

        boolean isIpn = paymentData.containsKey("transId");
        String rawSignature;
        if (isIpn) {
            Map<String, String> params = new LinkedHashMap<>();
            params.put("accessKey", accessKey);
            params.put("amount", paymentData.getOrDefault("amount", ""));
            params.put("extraData", paymentData.getOrDefault("extraData", ""));
            params.put("message", paymentData.getOrDefault("message", ""));
            params.put("orderId", paymentData.getOrDefault("orderId", ""));
            params.put("orderInfo", paymentData.getOrDefault("orderInfo", ""));
            params.put("orderType", paymentData.getOrDefault("orderType", ""));
            params.put("partnerCode", partnerCode);
            params.put("payType", paymentData.getOrDefault("payType", ""));
            params.put("requestId", paymentData.getOrDefault("requestId", ""));
            params.put("responseTime", paymentData.getOrDefault("responseTime", ""));
            params.put("resultCode", paymentData.getOrDefault("resultCode", ""));
            params.put("transId", paymentData.getOrDefault("transId", ""));

            rawSignature = buildSignatureString(params);
        } else {
            rawSignature = buildSignatureString(Map.of(
                    "accessKey", accessKey,
                    "amount", paymentData.getOrDefault("amount", ""),
                    "message", paymentData.getOrDefault("message", ""),
                    "orderId", paymentData.getOrDefault("orderId", ""),
                    "partnerCode", partnerCode,
                    "payUrl", paymentData.getOrDefault("payUrl", ""),
                    "requestId", paymentData.getOrDefault("requestId", ""),
                    "responseTime", paymentData.getOrDefault("responseTime", ""),
                    "resultCode", paymentData.getOrDefault("resultCode", "")
            ));
        }

        String calculated = hmacSha256Hex(secretKey, rawSignature);
        boolean match = calculated.equalsIgnoreCase(signature);
        boolean success = match && resultCode == 0;

        return new PaymentVerificationResponse(success, paymentData.get("orderId"), null, null, null, null);
    }

    @Override
    public Object queryPaymentFromProvider(String paymentId, HttpServletRequest httpServletRequest) {
        if (!supported) {
            throw new UnsupportedOperationException("Momo is not supported yet");
        }
        return null;
    }

    @Override
    public RefundResponse refundPayment(String paymentId, HttpServletRequest httpServletRequest) {
        if (!supported) {
            throw new UnsupportedOperationException("Momo is not supported yet");
        }
        throw new UnsupportedOperationException("Refund is not supported by Momo");
    }

    @Override
    public PaymentProvider getPaymentProviderName() {
        return PaymentProvider.MOMO;
    }

    private String buildSignatureString(Map<String, String> fields) {
        return fields.entrySet().stream()
                .sorted(Map.Entry.comparingByKey())
                .map(e -> e.getKey() + "=" + e.getValue())
                .reduce((a, b) -> a + "&" + b)
                .orElse("");
    }

    private String hmacSha256Hex(String key, String data) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), "HmacSHA256"));
            byte[] bytes = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte b : bytes) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (Exception ex) {
            log.error("Failed to HMAC: {}", ex.getMessage());
            return "";
        }
    }

    private String sendJson(String url, Map<String, Object> body) {
        try (var client = HttpClients.createDefault()) {
            HttpPost post = new HttpPost(url);
            post.setHeader("Content-Type", "application/json");
            post.setEntity(new StringEntity(mapper.writeValueAsString(body), StandardCharsets.UTF_8));
            try (CloseableHttpResponse response = client.execute(post)) {
                return EntityUtils.toString(response.getEntity(), StandardCharsets.UTF_8);
            }
        } catch (Exception e) {
            throw new RuntimeException("Failed to send request to MoMo", e);
        }
    }
}

File: backend/payment/src/main/kotlin/com/bbmovie/payment/service/payment/momo/MomoAdapter.kt
package com.bbmovie.payment.service.payment.momo

import com.bbmovie.payment.entity.PaymentProvider
import com.bbmovie.payment.entity.PaymentStatus
import com.bbmovie.payment.service.payment.PaymentProviderAdapter
import com.bbmovie.payment.service.payment.dto.PaymentRequest
import com.bbmovie.payment.service.payment.dto.PaymentResponse
import com.bbmovie.payment.service.payment.dto.PaymentVerification
import com.bbmovie.payment.service.payment.dto.RefundResponse
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import jakarta.servlet.http.HttpServletRequest
import org.apache.http.client.methods.CloseableHttpResponse
import org.apache.http.client.methods.HttpPost
import org.apache.http.entity.StringEntity
import org.apache.http.impl.client.HttpClients
import org.apache.http.util.EntityUtils
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.nio.charset.StandardCharsets
import java.util.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

@Service("momoProvider")
class MomoAdapter : PaymentProviderAdapter {

    @Value("\${payment.momo.partner-code}")
    private lateinit var partnerCode: String

    @Value("\${payment.momo.access-key}")
    private lateinit var accessKey: String

    @Value("\${payment.momo.secret-key}")
    private lateinit var secretKey: String

    @Value("\${payment.momo.sandbox:true}")
    private var sandbox: Boolean = true

    @Value("\${payment.momo.redirect-url}")
    private lateinit var redirectUrl: String

    @Value("\${payment.momo.ipn-url}")
    private lateinit var ipnUrl: String

    private val log = LoggerFactory.getLogger(MomoAdapter::class.java)
    private val mapper: ObjectMapper = jacksonObjectMapper()

    override fun processPayment(request: PaymentRequest, httpServletRequest: HttpServletRequest): PaymentResponse {
        val amount = requireNotNull(request.amount) { "amount is required" }.toLong()
        val orderId = request.orderId ?: ("Partner_Transaction_ID_" + System.currentTimeMillis())
        val requestId = "Request_ID_" + System.currentTimeMillis()
        val orderInfo = "Payment for order ${request.orderId ?: orderId}"
        val requestType = MomoConstraint.REQUEST_TYPE_CAPTURE_WALLET

        val extraDataJson = mapper.writeValueAsString(mapOf("skus" to ""))
        val extraData = Base64.getEncoder().encodeToString(extraDataJson.toByteArray(StandardCharsets.UTF_8))

        val body = mutableMapOf(
            "partnerCode" to partnerCode,
            "requestId" to requestId,
            "amount" to amount,
            "orderId" to orderId,
            "orderInfo" to orderInfo,
            "redirectUrl" to redirectUrl,
            "ipnUrl" to ipnUrl,
            "requestType" to requestType,
            "extraData" to extraData
        )

        val rawSignature = buildSignatureString(
            mapOf(
                "accessKey" to accessKey,
                "amount" to amount.toString(),
                "extraData" to extraData,
                "ipnUrl" to ipnUrl,
                "orderId" to orderId,
                "orderInfo" to orderInfo,
                "partnerCode" to partnerCode,
                "redirectUrl" to redirectUrl,
                "requestId" to requestId,
                "requestType" to requestType
            )
        )

        val signature = hmacSha256Hex(secretKey, rawSignature)
        body["signature"] = signature

        val responseBody = sendJson(
            MomoConstraint.CREATE_URL_TEST.takeIf { sandbox } ?: MomoConstraint.CREATE_URL_PROD, body
        )
        val response: Map<String, Any?> = mapper.readValue(responseBody)
        val resultCode = (response["resultCode"] as? Number)?.toInt() ?: -1
        val payUrl = response["payUrl"]?.toString()

        val status = if (resultCode == 0) PaymentStatus.PENDING else PaymentStatus.FAILED
        return PaymentResponse(
            transactionId = orderId,
            status = status,
            providerReference = payUrl
        )
    }

    override fun verifyPayment(
        paymentData: MutableMap<String, String>,
        httpServletRequest: HttpServletRequest
    ): PaymentVerification {
        val signature = paymentData["signature"]
        val resultCode = paymentData["resultCode"]?.toIntOrNull()
        if (signature.isNullOrBlank() || resultCode == null) {
            return PaymentVerification(false, null)
        }

        val isIpn = paymentData.containsKey("transId")

        val rawSignature = if (isIpn) {
            // IPN signature per docs
            buildSignatureString(
                mapOf(
                    "accessKey" to accessKey,
                    "amount" to (paymentData["amount"] ?: ""),
                    "extraData" to (paymentData["extraData"] ?: ""),
                    "message" to (paymentData["message"] ?: ""),
                    "orderId" to (paymentData["orderId"] ?: ""),
                    "orderInfo" to (paymentData["orderInfo"] ?: ""),
                    "orderType" to (paymentData["orderType"] ?: ""),
                    "partnerCode" to partnerCode,
                    "payType" to (paymentData["payType"] ?: ""),
                    "requestId" to (paymentData["requestId"] ?: ""),
                    "responseTime" to (paymentData["responseTime"] ?: ""),
                    "resultCode" to (paymentData["resultCode"] ?: ""),
                    "transId" to (paymentData["transId"] ?: "")
                )
            )
        } else {
            // Redirect/response signature per docs
            buildSignatureString(
                mapOf(
                    "accessKey" to accessKey,
                    "amount" to (paymentData["amount"] ?: ""),
                    "message" to (paymentData["message"] ?: ""),
                    "orderId" to (paymentData["orderId"] ?: ""),
                    "partnerCode" to partnerCode,
                    "payUrl" to (paymentData["payUrl"] ?: ""),
                    "requestId" to (paymentData["requestId"] ?: ""),
                    "responseTime" to (paymentData["responseTime"] ?: ""),
                    "resultCode" to (paymentData["resultCode"] ?: "")
                )
            )
        }

        val calculated = hmacSha256Hex(secretKey, rawSignature)
        val match = calculated.equals(signature, ignoreCase = true)
        val success = match && resultCode == 0
        val orderId = paymentData["orderId"]
        return PaymentVerification(success, orderId)
    }

    override fun refundPayment(paymentId: String, httpServletRequest: HttpServletRequest): RefundResponse {
        throw UnsupportedOperationException("Refund is not supported by Momo")
    }

    override fun getPaymentProviderName(): PaymentProvider = PaymentProvider.MOMO

    private fun buildSignatureString(fields: Map<String, String>): String {
        // MoMo requires keys sorted from a-z and joined with &
        return fields.toSortedMap().entries.joinToString("&") { (k, v) -> "$k=$v" }
    }

    private fun hmacSha256Hex(key: String, data: String): String {
        return try {
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(SecretKeySpec(key.toByteArray(StandardCharsets.UTF_8), "HmacSHA256"))
            val bytes = mac.doFinal(data.toByteArray(StandardCharsets.UTF_8))
            bytes.joinToString("") { "%02x".format(it) }
        } catch (ex: Exception) {
            log.error("Failed to HMAC: ${ex.message}")
            ""
        }
    }

    private fun sendJson(url: String, body: Map<String, Any?>): String {
        val client = HttpClients.createDefault()
        val post = HttpPost(url)
        post.setHeader("Content-Type", "application/json")
        post.entity = StringEntity(mapper.writeValueAsString(body), StandardCharsets.UTF_8)
        return client.use {
            val response: CloseableHttpResponse = it.execute(post)
            EntityUtils.toString(response.entity, StandardCharsets.UTF_8)
        }
    }
}


